Trigger route callbacks in set_outlets

ajjahn opened the pull request:
Storing the previous path (much like it used to store the old_targets), allows the routing callbacks to be triggered within `set_outlets` again. This brings back the callbacks on load, etc. 

Probably could use some refactoring though. `targets_for_path` gets some redundant calls.

jgaskins commented:
This looks good, but I think there may be an issue with it off the top of my head. The URL is global mutable state, so tracking its previous state on an instance can cause them to become unsynchronized. Most navigation happens via the `Link` component (which uses [the class method](https://github.com/clearwater-rb/clearwater/blob/master/opal/clearwater/link.rb#L51), but it's not the only source of navigation. If two different router instances (or an instance and the class) trigger navigation, the wrong callbacks will get triggered on the second one.

This is the reason I tried iterating over individual routers. Running callbacks on first render might be as simple as just iterating over the targets and explicitly calling `on_route_to`. As a quick fix, it might be doable in `Application#render`.

Alternatively, we could try tracking previous path on the Router class itself since it's a singleton. I'm not sure of any drawbacks to that, but there's a possibility of the same problem if someone subclasses Router (for example, to implement hash-based routing) while also using the canonical Clearwater::Router.

ajjahn commented:
Yep, I wondered about the implications of the instance variable for the previous path, but decided to try it out as a quick fix. Since the last implementation cached and instance variable of the previous targets, I figured it wouldn't be any worse off.

I think storing the previous path as a class instance variable is a decent stop gap for now. Ultimately there is probably another object that could be extracted to deal with resolving changes in the path. Possibly, an object that wraps/delegates to the `window.history` object.

jgaskins commented:
I just noticed, this PR removes the calling of callbacks on navigation via the `Router.navigate_to` class method.

ajjahn commented:
Yes, because `navigate_to` renders the apps, which will call `set_outlets`, where the callbacks are invoked. I've been using this branch in development on my app and it's working quite well.

jgaskins commented:
Ah, good point.

ajjahn commented:
Okay, I moved the previous path to the class level.

Also, this fixes a bug where routing callbacks were not being triggered on targets nested below dynamic segments.

Also, this adds specs covering routing callbacks/gets the specs passing. I split them up into discrete scenarios to make it easier to reason about them. Sorry if it's too verbose.

jgaskins commented:
Sweet, I have very limited internet at the moment so I will have a proper look when I get home Sunday night!

jgaskins commented:
This looks good. I like all the additional tests, too. 👍
